File: frontend/src/lib/personalization.ts
import { apiFetch } from "@/lib/api"

export type ExplanationDepth = "concise" | "balanced" | "detailed"
export type ExplanationStyle = "direct" | "step_by_step" | "socratic" | "example_driven"

export interface LearningPreferences {
  id: string | null
  user_id: string
  explanation_depth: ExplanationDepth | null
  explanation_style: ExplanationStyle | null
  created_at: string | null
  updated_at: string | null
}

export interface PreferenceSuggestion {
  id: string
  preference_key: "explanation_depth" | "explanation_style"
  suggested_value: string
  signal_type: string
  rationale: string
  status: "pending" | "accepted" | "rejected"
  created_at: string
  updated_at: string
}

export interface Recommendation {
  action: "review_topic" | "take_quiz" | "practice_challenge"
  priority: "high" | "medium" | "low"
  topic: string
  url: string
  rationale: string
}

export type LearningPreferenceUpdate = Partial<
  Pick<LearningPreferences, "explanation_depth" | "explanation_style">
>

export function fetchLearningPreferences(): Promise<LearningPreferences> {
  return apiFetch<LearningPreferences>("/users/me/learning-preferences")
}

export function updateLearningPreferences(
  update: LearningPreferenceUpdate,
): Promise<LearningPreferences> {
  return apiFetch<LearningPreferences>("/users/me/learning-preferences", {
    method: "PATCH",
    body: JSON.stringify(update),
  })
}

export function fetchPreferenceSuggestions(): Promise<PreferenceSuggestion[]> {
  return apiFetch<PreferenceSuggestion[]>("/users/me/learning-preference-suggestions")
}

export function refreshPreferenceSuggestions(): Promise<PreferenceSuggestion[]> {
  return apiFetch<PreferenceSuggestion[]>("/users/me/learning-preference-suggestions/refresh", {
    method: "POST",
  })
}

export function resolvePreferenceSuggestion(
  suggestionId: string,
  resolution: "accept" | "reject",
): Promise<PreferenceSuggestion> {
  return apiFetch<PreferenceSuggestion>(
    `/users/me/learning-preference-suggestions/${suggestionId}/${resolution}`,
    { method: "POST" },
  )
}

export function fetchRecommendations(notebookId: string): Promise<Recommendation[]> {
  return apiFetch<Recommendation[]>(`/notebooks/${notebookId}/recommendations`)
}

File: backend/app/schemas/personalization.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

ExplanationDepth = Literal["concise", "balanced", "detailed"]
ExplanationStyle = Literal["direct", "step_by_step", "socratic", "example_driven"]
SuggestionStatus = Literal["pending", "accepted", "rejected"]
RecommendationPriority = Literal["high", "medium", "low"]
RecommendationAction = Literal["review_topic", "take_quiz", "practice_challenge"]


class LearningPreferenceUpdate(BaseModel):
    explanation_depth: ExplanationDepth | None = None
    explanation_style: ExplanationStyle | None = None


class LearningPreferenceRead(LearningPreferenceUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID | None = None
    user_id: uuid.UUID
    created_at: datetime | None = None
    updated_at: datetime | None = None


class LearningPreferenceSuggestionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    preference_key: Literal["explanation_depth", "explanation_style"]
    suggested_value: str
    signal_type: str
    rationale: str
    status: SuggestionStatus
    created_at: datetime
    updated_at: datetime


class RecommendationRead(BaseModel):
    action: RecommendationAction
    priority: RecommendationPriority
    topic: str
    url: str
    rationale: str
